src/ray/core_worker/context.cc

```diff
@@ -303,15 +307,25 @@ void WorkerContext::SetCurrentTask(const TaskSpecification &task_spec) {
   } else {
     RAY_CHECK(false);
   }
+
   if (task_spec.IsNormalTask() || task_spec.IsActorCreationTask()) {
-    // TODO(architkulkarni): Once workers are cached by runtime env, we should
-    // only set runtime_env_ once and then RAY_CHECK that we
-    // never see a new one.
-    runtime_env_info_ = std::make_shared<rpc::RuntimeEnvInfo>();
-    *runtime_env_info_ = task_spec.RuntimeEnvInfo();
-    if (!IsRuntimeEnvEmpty(runtime_env_info_->serialized_runtime_env())) {
-      runtime_env_ = std::make_shared<nlohmann::json>();
-      *runtime_env_ = nlohmann::json::parse(runtime_env_info_->serialized_runtime_env());
+    if (runtime_env_info_ == nullptr) {
+      runtime_env_info_ = std::make_shared<rpc::RuntimeEnvInfo>();
+      *runtime_env_info_ = task_spec.RuntimeEnvInfo();
+    }
+
+    if (!IsRuntimeEnvEmpty(serialized_runtime_env)) {
+      // All threads are requesting for the same parsed json result, so ok to place in
+      // critical section.
+      if (!serialized_runtime_env_.empty()) {
+        RAY_CHECK_EQ(serialized_runtime_env_, serialized_runtime_env)
```

Comment:
As discussed, we should always do the check as long as runtime_env_info_ is null (first time assignment)

Reply:
Yeah makes sense, add check on invariants